src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNMethod.java

```diff
@@ -0,0 +1,271 @@
+     * Generate extra parameters that do not go in the method string
+     *
+     * @param knnMethodContext from where parameters are retrieved
+     * @return map of extra parameters
+     */
+    public Map<String, Object> generateExtraParameterMap(KNNMethodContext knnMethodContext) {
+        return methodComponent.generateExtraParameterMap(knnMethodContext.getMethodComponent().getParameters());
+    }
+
+    /**
+     * MethodComponent defines the structure of an individual component that can make up an index
+     */
+    public static class MethodComponent {
+
+        /**
+         * Constructor
+         * @param name name of component
+         * @param parameters parameters that the component can take
+         */
+        public MethodComponent(String name, Map<String, Parameter<?>> parameters) {
+            this.name = name;
+            this.parameters = parameters;
+        }
+
+        protected String name;
+        private Map<String, Parameter<?>> parameters;
+
+        /**
+         * Get the name of the component
+         *
+         * @return name
+         */
+        public String getName() {
+            return name;
+        }
+
+        /**
+         * Get the parameters for the component
+         *
+         * @return parameters
+         */
+        public Map<String, Parameter<?>> getParameters() {
+            return parameters;
+        }
+
+        /**
+         * Generate extra parameters for this component that are not contained in method string
+         *
+         * @return map of extra parameters
+         */
+        public Map<String, Object> generateExtraParameterMap(Map<String, Object> inParameters) {
+            if (inParameters == null) {
+                return Collections.emptyMap();
+            }
+            return inParameters.entrySet().stream().filter(v -> !parameters.get(v.getKey()).isInMethodString()).collect(
+                    Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+        }
+
+        /**
+         * Validate that the methodComponentContext is a valid configuration for this methodComponent
+         *
+         * @param methodComponentContext to be validated
+         */
+        public void validate(KNNMethodContext.MethodComponentContext methodComponentContext) {
+            Map<String, Object> providedParameters = methodComponentContext.getParameters();
+            if (providedParameters != null) {
+                for (Map.Entry<String, Object> parameter : providedParameters.entrySet()) {
+                    if (!parameters.containsKey(parameter.getKey())) {
+                        throw new ValidationException();
+                    }
+
+                    parameters.get(parameter.getKey()).validate(parameter.getValue());
+                }
+            }
+        }
+    }
+
+    /**
+     * Parameter that can be set for a method component
+     *
+     * @param <T> Type parameter takes
+     */
+    public static abstract class Parameter<T> {
```

Comment:
if it is public, may be can extract it into new file. What do you think?

Reply:
This may not be required to be public. Let me think about it.

Reply:
Kept it public and moved it out of the file